deltacat/exceptions.py

```diff
@@ -1,14 +1,144 @@
+    PYMEMCACHED_PUT_OBJECT_ERROR = "10170"
+
+    # Storage Error code from 10300 to 10399
+    GENERAL_STORAGE_ERROR = "10300"
+    STORAGE_CONCURRENT_MODIFICATION_ERROR = "10301"
+
+    # Throttling Error code from 10500 to 10599
+    GENERAL_THROTTLING_ERROR = "10500"
+    RETRYABLE_UPLOAD_TABLE_ERROR = "10501"
+    RETRYABLE_DOWNLOAD_TABLE_ERROR = "10502"
+
+    # Validation Error code from 10700 to 10799
+    GENERAL_VALIDATION_ERROR = "10700"
+    CONTENT_TYPE_VALIDATION_ERROR = "10701"
 
 
 class NonRetryableError(Exception):
-    pass
+    is_retryable = False
+
+
+class RetryableError(Exception):
+    is_retryable = True
+
+
+# >>> example: raise DependencyRayError(ray_task="x")
+#
+# >>> __main__.DependencyRayError: Error Code: 10101. Is Retryable Error: False. A dependency Ray error occurred, during ray_task: x
+class DependencyRayError(DeltaCatError):
+    error_code = DeltaCATErrorCodes.DEPENDENCY_RAY_ERROR.value
+
+
+class DependencyRayWorkerDiedError(DependencyRayError):
+    error_code = DeltaCATErrorCodes.DEPENDENCY_RAY_WORKER_DIED_ERROR.value
+
+
+class DependencyRayOutOfMemoryError(DependencyRayError):
+    error_code = DeltaCATErrorCodes.DEPENDENCY_RAY_OUT_OF_MEMORY_ERROR.value
+
+
+class DependencyRayRuntimeSetupError(DependencyRayError, RetryableError):
+    error_code = DeltaCATErrorCodes.DEPENDENCY_RAY_RUNTIME_SETUP_ERROR.value
+
+
+class DependencyPyarrowError(DeltaCatError, NonRetryableError):
+    error_code = DeltaCATErrorCodes.DEPENDENCY_PYARROW_ERROR.value
+
+
+class DependencyPyarrowInvalidError(DeltaCatError, NonRetryableError):
+    error_code = DeltaCATErrorCodes.DEPENDENCY_PYARROW_INVALID_ERROR.value
+
+
+class DependencyPyarrowCapacityError(DeltaCatError, NonRetryableError):
+    error_code = DeltaCATErrorCodes.DEPENDENCY_PYARROW_CAPACITY_ERROR.value
+
+
+class PymemcachedPutObjectError(DeltaCatError, RetryableError):
+    error_code = DeltaCATErrorCodes.PYMEMCACHED_PUT_OBJECT_ERROR.value
+
+
+class GeneralValidationError(DeltaCatError):
+    error_code = DeltaCATErrorCodes.GENERAL_VALIDATION_ERROR.value
+
+
+class ContentTypeValidationError(GeneralValidationError, NonRetryableError):
+    error_code = DeltaCATErrorCodes.CONTENT_TYPE_VALIDATION_ERROR.value
+
+
+class GeneralStorageError(DeltaCatError):
+    error_code = DeltaCATErrorCodes.GENERAL_STORAGE_ERROR.value
+
+
+class StorageConcurrentModificationError(GeneralStorageError, NonRetryableError):
+    error_code = DeltaCATErrorCodes.STORAGE_CONCURRENT_MODIFICATION_ERROR.value
+
+
+class DependencyBotocoreError(DeltaCatError):
+    error_code = DeltaCATErrorCodes.DEPENDENCY_BOTOCORE_ERROR.value
+
+
+class DownloadTableError(DependencyBotocoreError, NonRetryableError):
+    error_code = DeltaCATErrorCodes.DOWNLOAD_TABLE_ERROR.value
+
+
+class UploadTableError(DependencyBotocoreError, NonRetryableError):
+    error_code = DeltaCATErrorCodes.UPLOAD_TABLE_ERROR.value
+
+
+class GeneralThrottlingError(DeltaCatError, RetryableError):
```

Comment:
I don't see Daft related error in here which probably must be catched from our `read_file`.

Reply:
Could you link which specific Daft error you are talking about? I didn't see the error in our oncall handoff doc.

Reply:
https://github.com/Eventual-Inc/Daft/blob/39610c71ca2b40aebc2b43d4497af5ae3398d4dc/src/common/error/src/error.rs#L98

Reply:
Do you mean we just catch all DaftError and cast to a DeltaCAT error?

Reply:
Yes, like other errors you have done, cast it to known errors and valid error code.